src/main/java/htsjdk/samtools/util/IOUtil.java

```diff
@@ -1153,4 +1157,96 @@ public static Path getPath(String uriString) throws IOException {
     public static Path addExtension(Path path, String extension) {
         return path.resolveSibling(path.getFileName() + extension);
     }
+
+    /**
+     * Checks if the provided path is block-compressed.
+     *
+     * <p>Note that using {@code checkExtension=true} would avoid the cost of opening the file, but
+     * if {@link #hasBlockCompressedExtension(String)} returns {@code false} this would not detect
+     * block-compressed files such BAM.
+     *
+     * @param path file to check if it is block-compressed.
+     * @param checkExtension if {@code true}, checks the extension before opening the file.
+     * @return {@code true} if the file is block-compressed; {@code false} otherwise.
+     * @throws IOException if there is an I/O error.
+     */
+    public static boolean isBlockCompressed(final Path path, final boolean checkExtension) throws IOException {
+        if (checkExtension && !hasBlockCompressedExtension(path.toString())) {
+            return false;
+        }
+        try (final InputStream stream = new BufferedInputStream(Files.newInputStream(path), Math.max(Defaults.BUFFER_SIZE, BlockCompressedStreamConstants.MAX_COMPRESSED_BLOCK_SIZE))) {
+            return BlockCompressedInputStream.isValidFile(stream);
+        }
+    }
+
+    /**
+     * Checks if the provided path is block-compressed (including extension).
+     *
+     * <p>Note that block-compressed files with extensions not in {@link #BLOCK_COMPRESSED_EXTENSIONS}
+     * would return {@code false} with this method (e.g., BAM files) even if they are block-compressed.
+     * Use {@link #isBlockCompressed(Path, boolean)} if this is not expected.
+     *
+     * @param path file to check if it is block-compressed.
+     * @return {@code true} if the file is block-compressed; {@code false} otherwise.
+     * @throws IOException if there is an I/O error.
+     */
+    public static boolean isBlockCompressed(final Path path) throws IOException {
```

Comment:
Let's add `.bam` and maybe `.bcf` to the list of block compressed files so the extension check will work with those files.

Reply:
I agree with the default behavior but I am opposed to add `.bam` and `bcf` to the list of block-compressed extensions. The reason: block-compressed extensions are used in the wild (my own projects and many more) as a way to identify compressed plain text file  (gzip/bgzip) to read/write in the proper format (usually `.gz` or `.bgz` in my case). In addition, there are already methods to check if a file is a `.bam` or a `.bcf`...

Reply:
Revered the behavior in next commit

Reply:
I'm not sure how I feel about adding .bam and .bcf.  I think it would make sense, because they are block compressed, but they are sort of a special case because it's integral to the file type rather than an extension on it.  (although in the bcf case that's not quite true since htsjdk doesn't block compress it by default).  

What would the harm be in including them in the list?  I assume you still need an additional step to identify the actual type of the compressed file.

Reply:
I'm not sure I totally agree with you about the problems with adding bam and bcf to the list, but I don't feel strongly and I'm happy to leave this PR as is.  

About `.bgz`, I think it was a mistake to use `.gz` as the default extension for block gzipped file.  I suspect there are issues with extension checks that explicitly look for `.gz` scattered throughout the various codebases though.  I have no idea for instance if htslib supports reading things with the  `.bgz` extension.  I would like to move towards making bgz the standard output in the future, and we can start by making sure that htsjdk and downstream tools deal correctly with `.bgz` extensions.  

I don't think we'll ever be able to drop `gz` support as a block compressed input since it's been the standard so long, but we can at least start moving towards a more sensible future.